src/swupd_activator.cpp

```diff
@@ -0,0 +1,262 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright (C) 2022 YADRO.
+ */
+
+#include "dbus.hpp"
+
+#include <getopt.h>
+
+#include <filesystem>
+#include <set>
+
+namespace fs = std::filesystem;
+
+struct Version
+{
+    std::string version;
+    std::string status;
+    std::set<std::string> targets;
+};
+
+bool printVersionsList()
+{
+    auto bus = sdbusplus::bus::new_default();
+
+    SubTreeType objects;
+    auto getObjects =
+        bus.new_method_call(dbus::mapper::busName, dbus::mapper::path,
+                            dbus::mapper::interface, dbus::mapper::subtree);
+    getObjects.append(
+        dbus::software::path, 0,
+        std::array<std::string, 1>{dbus::software::activationIface});
+
+    try
+    {
+        bus.call(getObjects).read(objects);
+    }
+    catch (const sdbusplus::exception::SdBusError& ex)
+    {
+        fprintf(stderr, "Error while calling GetSubTree: %s\n", ex.what());
+        return false;
+    }
+
+    std::map<std::string, Version> versions;
+    for (const auto& [path, objDict] : objects)
+    {
+        const std::string& owner = objDict.begin()->first;
+        const Interfaces& ifaces = objDict.begin()->second;
+
+        auto it = std::find(ifaces.begin(), ifaces.end(),
+                            dbus::software::filepathIface);
+        if (it == ifaces.end())
+        {
+            continue;
+        }
+
+        std::string verId;
+        std::string ver;
+        std::string status;
+        std::set<std::string> targets;
+        std::string inventoryItem;
+
+        fs::path p(path);
+        verId = p.filename();
+        std::string parent = p.parent_path().filename();
+        if (parent != "software")
+        {
+            verId = parent;
+        }
+        DbusProperties versionData;
+        auto getProperties = bus.new_method_call(owner.c_str(), path.c_str(),
+                                                 dbus::properties::interface,
+                                                 dbus::properties::getAll);
+        getProperties.append(dbus::software::versionIface);
```

Comment:
I suggest to not restrict the properties by the version interface.
It allows to receive all of them in the one call.
In this way we lose info about the property interface, but they have unique names and it should be enough.